New AHBot implementation

canidae opened the pull request:
This is a near complete rewrite of the AuctionHouseBot code.
It mimics players killing and looting creatures, and gathering loot from gameobjects, skinning, disenchanting and fishing, without actually removing creatures & objects from the game.
The previous implementation did not consider drop chances of items, and several items were with the default settings never added to the AH as they were sold by vendors in limited supplies (notably herbs/flowers suffered from this).
The AHBot buyer could easily be tricked into buying items sold by vendors to more than 100 times the sell price, giving players an infinite source of income (although the AHBot buyer was disabled by
default).
This commit also stops sending players a notification about "sold" AH item when the AHBot merely bids on a player auctioned item.

### Proof
- None

### Issues
This change affects the following issues:
- https://github.com/cmangos/issues/issues/1762
- https://github.com/cmangos/issues/issues/1541

### How2Test
- ".ahbot status" shows they current total amount of AHBot-placed items in each AH.
- ".ahbot reload" reloads the configuration file.
- ".ahbot rebuild" will fill the AH with items, with an approximation of how many items there should be in the AH at a typical equilibrium.

### Todo / Checklist
- [x] There may be items that shouldn't be added to the AH which still will end up there.
- [x] Item sources may need to be better balanced in the default configuration, but this is an insignificant modification that can be done at a later time.
- [x] Currently the code will force a buyout price for items sold by vendor to 4 times the item sell value. This is to prevent a possible loophole where green/blue/purple items (which by default has buyout value set to 8/16/32 times the item sell value) can be bought by vendors and sold at the AH for a significant income. This may not be an issue and in that case the code for this check can be removed, but the check is hardly computationally expensive.
- [x] Linked AHs should be tested. Item count is likely to be tripled, I don't know what should be the right behavior here (divide item count by 3 or let it be tripled).

canidae commented:
I'm still testing this new AHBot on a live server (with 5-6 active people) and may suggests a couple more changes, but this will take time so I'm opening the PR now for review. It can be merged now if so desired, I'll just create a new PR if I discover some additional changes that should be done after this PR is merged.

ekg98 commented:
Will the buyout check be optional or controllable via conf. I have not looked at the code yet. On my phone so its hard. Some of us like it boosted a tad but still realistic.

canidae commented:
> Will the buyout check be optional or controllable via conf. I have not looked at the code yet. On my phone so its hard. Some of us like it boosted a tad but still realistic. Ps found it. Very cool addon.

Currently it's not configurable, but this is a good candidate for a configurable option. It's likely I'll look into this later.

canidae commented:
As a comment for those with servers having the current AHBot buyer enabled:
If there are no "Demonic Figurine" on the AH, buy a stack (5) from a reagent vendor (costing you no more than 5 gold). Put it up for auction at 146 gold (set both bid and buyout to this value). It's likely it'll be sold within a couple hours.
You can do this with any item not available on the AH (even gray items). Rule of thumb is to set bid and buyout to 100 times the item sell value (although you can go even further).

canidae commented:
Made "vendor multiplier" and probability of AHBot buyer "visiting" the AH configurable options.

A current issue I'm looking into is that buying magic items, disenchanting them and selling the dust/essence/shards on AH is a net positive endeavor. Making it lucrative to buy all magic items from AH, disenchanting them and selling the resulting mats.

canidae commented:
Made "vendor multiplier" and probability of AHBot buyer "visiting" the AH configurable options.

A current issue I'm looking into is that buying magic items, disenchanting them and selling the dust/essence/shards on AH is a net positive endeavor. Making it lucrative to buy all magic items from AH, disenchanting them and selling the resulting mats.

Daedalus-code commented:
Max buy a / day somewhere?
Max spending a day divided into several buys.

Anything to stop this economic nightmare with this buyer "enabled" :P 
Sold a lvl 6 Shield for 200g bid.

canidae commented:
Fairly satisfied with the AHBot now. Recap for issues mentioned:
* Disenchanting & selling essence/shard/etc: Should no longer be an easy source of income. Enchanting mats don't have a vendor sell value, so I used vendor buy value instead divided by 4. However, magic (green, blue, purple) items cost 5 times as much from vendors. Hard to explain, but short story is that enchanting mats had too high value relative to magic armor/weapons.
* AH item balance: It is a subjective opinion how many and what kind of items there should be, but with the current configuration there will be about 3000 items at the AH, with a bit of everything.
* I've not been able to find any items that shouldn't be in the AH.
* I'm not aware of any way to abuse the AHBot into a (practical) source of infinite income (you can configure it in a way that makes it possible, though).
* I've not done anything with linked AHs. Those who set up a server with linked AHs should consider reducing Chance.Sell and Chance.Buy config to 1/3 of their original value.

cala commented:
Greetings,

That's close to a complete (and much needed) rewrite of AHBot. Thank you very much.

I/we will need some time to review it, especially as this is a custom/optional part of the core and we have currently our hands full. But this looks very nice and promising.

canidae commented:
Absolutely. Just let me know if you have any questions or comments.

I will continue looking into AH item balance, but as noted this is very much a subjective opinion. One change I'm considering is reducing `AuctionHouseBot.Chance.Buy` from 100 to e.g. 20 (and at the same time increase `AuctionHouseBot.Loot.*` values to keep about the same amount of items at the AH). The primary goal behind this idea is that you can easily configure "more of everything" without having to modify the `AuctionHouseBot.Loot.*` config (which is a bit complex).

canidae commented:
Regretably I've had some strange segfaults lately, and it seems like my implementation of AHBot is the cause, but I'm confused as of why. I managed to attach gdb to the server and catch the backtrace, and it is as follows:
```
Thread 7 "mangosd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ff720d90700 (LWP 1403)]
0x00007ff732770f63 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ff732770f63 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000055af77a34ae1 in std::_Rb_tree_const_iterator<std::pair<unsigned int const, AuctionEntry*> >::operator++ (this=0x7ff720d8fc98) at /usr/include/c++/7/bits/stl_tree.h:366
#2  0x000055af77a3424f in AuctionHouseBot::Update (this=0x55af7a96a740) at /cmangos/mangos-classic/src/game/AuctionHouseBot/AuctionHouseBot.cpp:296
#3  0x000055af77e522b7 in World::Update (this=0x55af7a6bfd40, diff=39) at /cmangos/mangos-classic/src/game/World/World.cpp:1408
#4  0x000055af779f6fc7 in WorldRunnable::run (this=0x55af8750d610) at /cmangos/mangos-classic/src/mangosd/WorldRunnable.cpp:59
#5  0x000055af77a2f2cf in MaNGOS::Thread::ThreadTask (param=0x55af8750d610) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#6  0x000055af77a2f85b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x55af8750a350: 0x55af77a2f2a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x55af8750a348: 0x55af8750d610) at /usr/include/c++/7/bits/invoke.h:60
#7  0x000055af77a2f54c in std::__invoke<void (*)(void*), void*> (__fn=@0x55af8750a350: 0x55af77a2f2a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x55af8750a348: 0x55af8750d610) at /usr/include/c++/7/bits/invoke.h:95
#8  0x000055af77a2fb11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x55af8750a348) at /usr/include/c++/7/thread:234
#9  0x000055af77a2fab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x55af8750a348) at /usr/include/c++/7/thread:243
#10 0x000055af77a2fa82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x55af8750a340) at /usr/include/c++/7/thread:186
#11 0x00007ff73278466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007ff7332856db in start_thread (arg=0x7ff720d90700) at pthread_create.c:463
#13 0x00007ff731e4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
The strange thing is that AuctionHouseBot.cpp:296 is the *second* line here:
```
        AuctionHouseObject::AuctionEntryMapBounds bounds = auctionHouse->GetAuctionsBounds();
        for (AuctionHouseObject::AuctionEntryMap::const_iterator itr = bounds.first; itr != bounds.second; ++itr) {
            AuctionEntry* auction = itr->second;
            ...
        }
```
This is not really "new" code, the existing AHBot does the same thing. C++ isn't my primary language, nor am I well versed in gdb, so if anyone have any ideas I'd much appreciate it. @cala? (sorry for hilighting, you've been too helpful)

evil-at-wow commented:
> std::_Rb_tree_const_iterator<std::pair<unsigned int const, AuctionEntry*> >::operator++

This is incrementing an iterator (the _Rb_tree no doubt refers to a red-black tree, a data structure typically used to implement a set/map). So it's normal that this is the _second_ line of the code block you refer to; it's the ```++iter``` that's segfaulting.

It looks like the iterator is invalidated while iterating over the map (that for-loop you have there is clearly iterating over the map). I don't immediately see where, but a typical case is erasing the item the iterator is pointing to. Another case could be the map changes 'behind your back', say when another thread does something with it (std::map is not thread-safe). If you get another crash like this, can you run ```info threads``` and ```thread apply all bt``` in gdb please? That will give you some info about other threads, and a backtrace (bt) of all the threads to see what other threads are doing at that time.

lucky until now.

canidae commented:
> It looks like the iterator is invalidated while iterating over the map (that for-loop you have there is clearly iterating over the map). I don't immediately see where, but a typical case is erasing the item the iterator is pointing to. Another case could be the map changes 'behind your back', say when another thread does something with it (std::map is not thread-safe). If you get another crash like this, can you run `info threads` and `thread apply all bt` in gdb please? That will give you some info about other threads, and a backtrace (bt) of all the threads to see what other threads are doing at that time.

Thanks, I will try this. It is indeed iterating over a map (auction items). It's quite possible another thread is removing items from that map as items expire. I was under the impression that the iterating was thread safe, but that doesn't seem to be the case.
Strange thing is that server pretty much exclusively crash when a certain character is logged on. Server is fine for hours, even days, but when one character is logged on it'll crash within the hour. Possibly just a coincidence, but having that character log in is still the only way to reliably cause a segfault.

canidae commented:
New crash. Again server is stable until a certain character joins. I'm utterly confused as of what's going on:
```
Thread 7 "mangosd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f64eadad700 (LWP 1403)]
0x00007f64fc770f63 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007f64fc770f63 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00005636ba287ae1 in std::_Rb_tree_const_iterator<std::pair<unsigned int const, AuctionEntry*> >::operator++ (this=0x7f64eadacc98) at /usr/include/c++/7/bits/stl_tree.h:366
#2  0x00005636ba28724f in AuctionHouseBot::Update (this=0x5636bb9ec740) at /cmangos/mangos-classic/src/game/AuctionHouseBot/AuctionHouseBot.cpp:296
#3  0x00005636ba6a52b7 in World::Update (this=0x5636bb741d40, diff=62) at /cmangos/mangos-classic/src/game/World/World.cpp:1408
#4  0x00005636ba249fc7 in WorldRunnable::run (this=0x5636c81d6770) at /cmangos/mangos-classic/src/mangosd/WorldRunnable.cpp:59
#5  0x00005636ba2822cf in MaNGOS::Thread::ThreadTask (param=0x5636c81d6770) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#6  0x00005636ba28285b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x5636c81d34b0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636c81d34a8: 0x5636c81d6770) at /usr/include/c++/7/bits/invoke.h:60
#7  0x00005636ba28254c in std::__invoke<void (*)(void*), void*> (__fn=@0x5636c81d34b0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636c81d34a8: 0x5636c81d6770) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005636ba282b11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x5636c81d34a8) at /usr/include/c++/7/thread:234
#9  0x00005636ba282ab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x5636c81d34a8) at /usr/include/c++/7/thread:243
#10 0x00005636ba282a82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x5636c81d34a0) at /usr/include/c++/7/thread:186
#11 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f64fd2856db in start_thread (arg=0x7f64eadad700) at pthread_create.c:463
#13 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) info threads
  Id   Target Id         Frame 
  1    Thread 0x7f64fdb83740 (LWP 1388) "mangosd" 0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7ffce640a3f0, remaining=0x7ffce640a3f0) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
  2    Thread 0x7f64fb909700 (LWP 1394) "mangosd" 0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64fb908d10, remaining=0x7f64fb908d10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
  3    Thread 0x7f64fb108700 (LWP 1397) "mangosd" 0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64fb107d10, remaining=0x7f64fb107d10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
  4    Thread 0x7f64fa907700 (LWP 1400) "mangosd" 0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64fa906d10, remaining=0x7f64fa906d10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
* 7    Thread 0x7f64eadad700 (LWP 1403) "mangosd" 0x00007f64fc770f63 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
  8    Thread 0x7f64eb5ae700 (LWP 1404) "mangosd" 0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64eb5adb90, remaining=0x7f64eb5adb90) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
  9    Thread 0x7f64e32ce700 (LWP 1405) "mangosd" 0x00007f64fbe41bb7 in epoll_wait (epfd=18, events=0x7f64e32cd640, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  10   Thread 0x7f64e2acd700 (LWP 1406) "mangosd" 0x00007f64fbe41bb7 in epoll_wait (epfd=14, events=0x7f64e2acc630, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  11   Thread 0x7f64e22cc700 (LWP 1407) "mangosd" 0x00007f64fbe41bb7 in epoll_wait (epfd=25, events=0x7f64e22cb640, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  12   Thread 0x7f64e1acb700 (LWP 1408) "mangosd" 0x00007f64fbe41bb7 in epoll_wait (epfd=21, events=0x7f64e1aca630, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
(gdb) thread apply all bt

Thread 12 (Thread 0x7f64e1acb700 (LWP 1408)):
#0  0x00007f64fbe41bb7 in epoll_wait (epfd=21, events=0x7f64e1aca630, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  0x00005636ba231a17 in boost::asio::detail::epoll_reactor::run (this=0x5636c81d8410, block=true, ops=...) at /usr/include/boost/asio/detail/impl/epoll_reactor.ipp:438
#2  0x00005636ba232f1d in boost::asio::detail::task_io_service::do_run_one (this=0x5636c81d82c0, lock=..., this_thread=..., ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:356
#3  0x00005636ba232a25 in boost::asio::detail::task_io_service::run (this=0x5636c81d82c0, ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:149
#4  0x00005636ba2332b0 in boost::asio::io_service::run (this=0x5636c81d8250) at /usr/include/boost/asio/impl/io_service.ipp:59
#5  0x00005636ba237071 in MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}::operator()() const (__closure=0x5636c81d8b28) at /cmangos/mangos-classic/src/shared/Network/Listener.hpp:82
#6  0x00005636ba23b847 in std::__invoke_impl<void, MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}>(std::__invoke_other, MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}&&) (__f=...) at /usr/include/c++/7/bits/invoke.h:60
#7  0x00005636ba2394a5 in std::__invoke<MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}>(std::__invoke_result&&, (MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}&&)...) (__fn=...) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005636ba24685a in std::thread::_Invoker<std::tuple<MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x5636c81d8b28) at /usr/include/c++/7/thread:234
#9  0x00005636ba24659a in std::thread::_Invoker<std::tuple<MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}> >::operator()() (this=0x5636c81d8b28) at /usr/include/c++/7/thread:243
#10 0x00005636ba246212 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<MaNGOS::Listener<RASocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}> > >::_M_run() (this=0x5636c81d8b20) at /usr/include/c++/7/thread:186
#11 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f64fd2856db in start_thread (arg=0x7f64e1acb700) at pthread_create.c:463
#13 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 11 (Thread 0x7f64e22cc700 (LWP 1407)):
#0  0x00007f64fbe41bb7 in epoll_wait (epfd=25, events=0x7f64e22cb640, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  0x00005636ba231a17 in boost::asio::detail::epoll_reactor::run (this=0x5636c81d8910, block=true, ops=...) at /usr/include/boost/asio/detail/impl/epoll_reactor.ipp:438
#2  0x00005636ba232f1d in boost::asio::detail::task_io_service::do_run_one (this=0x5636c81d8670, lock=..., this_thread=..., ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:356
#3  0x00005636ba232a25 in boost::asio::detail::task_io_service::run (this=0x5636c81d8670, ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:149
#4  0x00005636ba233301 in boost::asio::io_service::run (this=0x5636c81d8590, ec=...) at /usr/include/boost/asio/impl/io_service.ipp:66
#5  0x00005636ba238f34 in MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}::operator()() const (__closure=0x5636c81d8778) at /cmangos/mangos-classic/src/shared/Network/NetworkThread.hpp:47
#6  0x00005636ba23d0ef in std::__invoke_impl<void, MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}>(std::__invoke_other, MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}&&) (__f=...) at /usr/include/c++/7/bits/invoke.h:60
#7  0x00005636ba23b192 in std::__invoke<MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}>(std::__invoke_result&&, (MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}&&)...) (__fn=...) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005636ba246802 in std::thread::_Invoker<std::tuple<MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x5636c81d8778) at /usr/include/c++/7/thread:234
#9  0x00005636ba24650a in std::thread::_Invoker<std::tuple<MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}> >::operator()() (this=0x5636c81d8778) at /usr/include/c++/7/thread:243
#10 0x00005636ba2461d2 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<MaNGOS::NetworkThread<RASocket>::NetworkThread()::{lambda()#1}> > >::_M_run() (this=0x5636c81d8770) at /usr/include/c++/7/thread:186
#11 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f64fd2856db in start_thread (arg=0x7f64e22cc700) at pthread_create.c:463
#13 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 10 (Thread 0x7f64e2acd700 (LWP 1406)):
#0  0x00007f64fbe41bb7 in epoll_wait (epfd=14, events=0x7f64e2acc630, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  0x00005636ba231a17 in boost::asio::detail::epoll_reactor::run (this=0x5636c957d690, block=true, ops=...) at /usr/include/boost/asio/detail/impl/epoll_reactor.ipp:438
#2  0x00005636ba232f1d in boost::asio::detail::task_io_service::do_run_one (this=0x5636c81d7660, lock=..., this_thread=..., ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:356
#3  0x00005636ba232a25 in boost::asio::detail::task_io_service::run (this=0x5636c81d7660, ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:149
#4  0x00005636ba2332b0 in boost::asio::io_service::run (this=0x5636c81d75f0) at /usr/include/boost/asio/impl/io_service.ipp:59
#5  0x00005636ba236c83 in MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}::operator()() const (__closure=0x5636c81d80b8) at /cmangos/mangos-classic/src/shared/Network/Listener.hpp:82
#6  0x00005636ba23abf1 in std::__invoke_impl<void, MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}>(std::__invoke_other, MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}&&) (__f=...) at /usr/include/c++/7/bits/invoke.h:60
#7  0x00005636ba238aad in std::__invoke<MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}>(std::__invoke_result&&, (MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}&&)...) (__fn=...) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005636ba246886 in std::thread::_Invoker<std::tuple<MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x5636c81d80b8) at /usr/include/c++/7/thread:234
#9  0x00005636ba2465e2 in std::thread::_Invoker<std::tuple<MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}> >::operator()() (this=0x5636c81d80b8) at /usr/include/c++/7/thread:243
#10 0x00005636ba246232 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<MaNGOS::Listener<WorldSocket>::Listener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)::{lambda()#1}> > >::_M_run() (this=0x5636c81d80b0) at /usr/include/c++/7/thread:186
#11 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f64fd2856db in start_thread (arg=0x7f64e2acd700) at pthread_create.c:463
#13 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 9 (Thread 0x7f64e32ce700 (LWP 1405)):
#0  0x00007f64fbe41bb7 in epoll_wait (epfd=18, events=0x7f64e32cd640, maxevents=128, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  0x00005636ba231a17 in boost::asio::detail::epoll_reactor::run (this=0x5636c81d7f00, block=true, ops=...) at /usr/include/boost/asio/detail/impl/epoll_reactor.ipp:438
#2  0x00005636ba232f1d in boost::asio::detail::task_io_service::do_run_one (this=0x5636c81d7880, lock=..., this_thread=..., ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:356
#3  0x00005636ba232a25 in boost::asio::detail::task_io_service::run (this=0x5636c81d7880, ec=...) at /usr/include/boost/asio/detail/impl/task_io_service.ipp:149
#4  0x00005636ba233301 in boost::asio::io_service::run (this=0x5636c81d77a0, ec=...) at /usr/include/boost/asio/impl/io_service.ipp:66
#5  0x00005636ba238522 in MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}::operator()() const (__closure=0x5636c81d7988) at /cmangos/mangos-classic/src/shared/Network/NetworkThread.hpp:47
#6  0x00005636ba23c3ff in std::__invoke_impl<void, MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}>(std::__invoke_other, MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}&&) (__f=...) at /usr/include/c++/7/bits/invoke.h:60
#7  0x00005636ba23a500 in std::__invoke<MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}>(std::__invoke_result&&, (MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}&&)...) (__fn=...) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005636ba24682e in std::thread::_Invoker<std::tuple<MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x5636c81d7988) at /usr/include/c++/7/thread:234
#9  0x00005636ba246552 in std::thread::_Invoker<std::tuple<MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}> >::operator()() (this=0x5636c81d7988) at /usr/include/c++/7/thread:243
#10 0x00005636ba2461f2 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<MaNGOS::NetworkThread<WorldSocket>::NetworkThread()::{lambda()#1}> > >::_M_run() (this=0x5636c81d7980) at /usr/include/c++/7/thread:186
#11 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f64fd2856db in start_thread (arg=0x7f64e32ce700) at pthread_create.c:463
#13 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 8 (Thread 0x7f64eb5ae700 (LWP 1404)):
#0  0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64eb5adb90, remaining=0x7f64eb5adb90) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  0x00005636ba22038e in std::this_thread::sleep_for<long, std::ratio<1l, 1000000000l> > (__rtime=...) at /usr/include/c++/7/thread:373
#2  0x00005636ba21d1a5 in CliRunnable::run (this=0x5636c81d67e0) at /cmangos/mangos-classic/src/mangosd/CliRunnable.cpp:629
#3  0x00005636ba2822cf in MaNGOS::Thread::ThreadTask (param=0x5636c81d67e0) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#4  0x00005636ba28285b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x5636c81d75e0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636c81d75d8: 0x5636c81d67e0) at /usr/include/c++/7/bits/invoke.h:60
#5  0x00005636ba28254c in std::__invoke<void (*)(void*), void*> (__fn=@0x5636c81d75e0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636c81d75d8: 0x5636c81d67e0) at /usr/include/c++/7/bits/invoke.h:95
#6  0x00005636ba282b11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x5636c81d75d8) at /usr/include/c++/7/thread:234
#7  0x00005636ba282ab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x5636c81d75d8) at /usr/include/c++/7/thread:243
#8  0x00005636ba282a82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x5636c81d75d0) at /usr/include/c++/7/thread:186
#9  0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#10 0x00007f64fd2856db in start_thread (arg=0x7f64eb5ae700) at pthread_create.c:463
#11 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 7 (Thread 0x7f64eadad700 (LWP 1403)):
#0  0x00007f64fc770f63 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00005636ba287ae1 in std::_Rb_tree_const_iterator<std::pair<unsigned int const, AuctionEntry*> >::operator++ (this=0x7f64eadacc98) at /usr/include/c++/7/bits/stl_tree.h:366
#2  0x00005636ba28724f in AuctionHouseBot::Update (this=0x5636bb9ec740) at /cmangos/mangos-classic/src/game/AuctionHouseBot/AuctionHouseBot.cpp:296
#3  0x00005636ba6a52b7 in World::Update (this=0x5636bb741d40, diff=62) at /cmangos/mangos-classic/src/game/World/World.cpp:1408
#4  0x00005636ba249fc7 in WorldRunnable::run (this=0x5636c81d6770) at /cmangos/mangos-classic/src/mangosd/WorldRunnable.cpp:59
#5  0x00005636ba2822cf in MaNGOS::Thread::ThreadTask (param=0x5636c81d6770) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#6  0x00005636ba28285b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x5636c81d34b0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636c81d34a8: 0x5636c81d6770) at /usr/include/c++/7/bits/invoke.h:60
#7  0x00005636ba28254c in std::__invoke<void (*)(void*), void*> (__fn=@0x5636c81d34b0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636c81d34a8: 0x5636c81d6770) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005636ba282b11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x5636c81d34a8) at /usr/include/c++/7/thread:234
#9  0x00005636ba282ab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x5636c81d34a8) at /usr/include/c++/7/thread:243
#10 0x00005636ba282a82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x5636c81d34a0) at /usr/include/c++/7/thread:186
#11 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f64fd2856db in start_thread (arg=0x7f64eadad700) at pthread_create.c:463
#13 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 4 (Thread 0x7f64fa907700 (LWP 1400)):
#0  0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64fa906d10, remaining=0x7f64fa906d10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  0x00005636ba224cd7 in std::this_thread::sleep_for<long, std::ratio<1l, 1000l> > (__rtime=...) at /usr/include/c++/7/thread:373
#2  0x00005636ba282398 in MaNGOS::Thread::Sleep (msecs=10) at /cmangos/mangos-classic/src/shared/Threading.cpp:136
#3  0x00005636ba26238e in SqlDelayThread::run (this=0x5636bb741c90) at /cmangos/mangos-classic/src/shared/Database/SqlDelayThread.cpp:48
#4  0x00005636ba2822cf in MaNGOS::Thread::ThreadTask (param=0x5636bb741c90) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#5  0x00005636ba28285b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x5636bb7419d0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636bb7419c8: 0x5636bb741c90) at /usr/include/c++/7/bits/invoke.h:60
#6  0x00005636ba28254c in std::__invoke<void (*)(void*), void*> (__fn=@0x5636bb7419d0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636bb7419c8: 0x5636bb741c90) at /usr/include/c++/7/bits/invoke.h:95
#7  0x00005636ba282b11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x5636bb7419c8) at /usr/include/c++/7/thread:234
#8  0x00005636ba282ab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x5636bb7419c8) at /usr/include/c++/7/thread:243
#9  0x00005636ba282a82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x5636bb7419c0) at /usr/include/c++/7/thread:186
#10 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007f64fd2856db in start_thread (arg=0x7f64fa907700) at pthread_create.c:463
#12 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 3 (Thread 0x7f64fb108700 (LWP 1397)):
#0  0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64fb107d10, remaining=0x7f64fb107d10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  0x00005636ba224cd7 in std::this_thread::sleep_for<long, std::ratio<1l, 1000l> > (__rtime=...) at /usr/include/c++/7/thread:373
#2  0x00005636ba282398 in MaNGOS::Thread::Sleep (msecs=10) at /cmangos/mangos-classic/src/shared/Threading.cpp:136
#3  0x00005636ba26238e in SqlDelayThread::run (this=0x5636bb725950) at /cmangos/mangos-classic/src/shared/Database/SqlDelayThread.cpp:48
#4  0x00005636ba2822cf in MaNGOS::Thread::ThreadTask (param=0x5636bb725950) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#5  0x00005636ba28285b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x5636bb726ec0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636bb726eb8: 0x5636bb725950) at /usr/include/c++/7/bits/invoke.h:60
#6  0x00005636ba28254c in std::__invoke<void (*)(void*), void*> (__fn=@0x5636bb726ec0: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636bb726eb8: 0x5636bb725950) at /usr/include/c++/7/bits/invoke.h:95
#7  0x00005636ba282b11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x5636bb726eb8) at /usr/include/c++/7/thread:234
#8  0x00005636ba282ab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x5636bb726eb8) at /usr/include/c++/7/thread:243
#9  0x00005636ba282a82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x5636bb726eb0) at /usr/include/c++/7/thread:186
#10 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007f64fd2856db in start_thread (arg=0x7f64fb108700) at pthread_create.c:463
#12 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 2 (Thread 0x7f64fb909700 (LWP 1394)):
#0  0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7f64fb908d10, remaining=0x7f64fb908d10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  0x00005636ba224cd7 in std::this_thread::sleep_for<long, std::ratio<1l, 1000l> > (__rtime=...) at /usr/include/c++/7/thread:373
#2  0x00005636ba282398 in MaNGOS::Thread::Sleep (msecs=10) at /cmangos/mangos-classic/src/shared/Threading.cpp:136
#3  0x00005636ba26238e in SqlDelayThread::run (this=0x5636bb711df0) at /cmangos/mangos-classic/src/shared/Database/SqlDelayThread.cpp:48
#4  0x00005636ba2822cf in MaNGOS::Thread::ThreadTask (param=0x5636bb711df0) at /cmangos/mangos-classic/src/shared/Threading.cpp:84
#5  0x00005636ba28285b in std::__invoke_impl<void, void (*)(void*), void*> (__f=@0x5636bb725710: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636bb725708: 0x5636bb711df0) at /usr/include/c++/7/bits/invoke.h:60
#6  0x00005636ba28254c in std::__invoke<void (*)(void*), void*> (__fn=@0x5636bb725710: 0x5636ba2822a4 <MaNGOS::Thread::ThreadTask(void*)>, __args#0=@0x5636bb725708: 0x5636bb711df0) at /usr/include/c++/7/bits/invoke.h:95
#7  0x00005636ba282b11 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::_M_invoke<0ul, 1ul> (this=0x5636bb725708) at /usr/include/c++/7/thread:234
#8  0x00005636ba282ab2 in std::thread::_Invoker<std::tuple<void (*)(void*), void*> >::operator() (this=0x5636bb725708) at /usr/include/c++/7/thread:243
#9  0x00005636ba282a82 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(void*), void*> > >::_M_run (this=0x5636bb725700) at /usr/include/c++/7/thread:186
#10 0x00007f64fc78466f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007f64fd2856db in start_thread (arg=0x7f64fb909700) at pthread_create.c:463
#12 0x00007f64fbe4188f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 1 (Thread 0x7f64fdb83740 (LWP 1388)):
#0  0x00007f64fd28fc60 in __GI___nanosleep (requested_time=0x7ffce640a3f0, remaining=0x7ffce640a3f0) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  0x00005636ba2376cd in std::this_thread::sleep_for<long, std::ratio<1l, 1l> > (__rtime=...) at /usr/include/c++/7/thread:373
#2  0x00005636ba22e508 in Master::Run (this=0x5636bb70d840) at /cmangos/mangos-classic/src/mangosd/Master.cpp:236
#3  0x00005636ba226aaf in main (argc=1, argv=0x7ffce640a8b8) at /cmangos/mangos-classic/src/mangosd/Main.cpp:200
```

evil-at-wow commented:
No other threads are really doing anything at that very moment (except for sleeping basically, waiting to do something). So if some other thread was messing with the map behind your back, it's not visible here. And timing wise you really are unlucky then, so there must be something else we're both missing. 

It's annoying there's no more or less easy way to reproduce, otherwise I could try your branch here. 

That particular character is a curious observation. I could be doing something to trigger this issue, but what... Just to rule out wild ideas: is that someone you know or trust? Meaning it's very unlikely that this character is modifying his/her traffic / edit packets etc?

canidae commented:
> That particular character is a curious observation. I could be doing something to trigger this issue, but what... Just to rule out wild ideas: is that someone you know or trust? Meaning it's very unlikely that this character is modifying his/her traffic / edit packets etc?

This is a person I trust, and it's someone who've helped me debug this issue. However, I made the server log crashes more explicitly and it turns out that the server isn't only crashing when this player is online.
However, I think I now may have figured it out:
While iterating AH items the bot may buyout items. This immediately removes the item from the map, invalidating the iterator. The AHBot won't buyout its own items, but it may buy items bidded on by a player, or items placed on AH by a player. Pluss an element of randomness to when it decides to buy items and we'll end up with unpredictable crashes (I'm not entirely convinced it would always crash either, I've tested buyout several times without issues, but that was on a dev-server).

I'll look into fixing this within the next days. Thanks a lot for your comments, they've been really helpful!

evil-at-wow commented:
I'm glad to see you managed to find the root cause of this problem! And no problem, I'm always happy if I can contribute to stability and quality :smile:

canidae commented:
An update on status:
I'm doing some final tweaks on config now. Mostly satisfied, items seems fairly well balanced. Enchantment mats could be somewhat better, but it would require special treatment within the code so I'm thinking it's not worth it (arguably there should be more stuff of lower level and less of higher level, and not so uniformly distributed as it is now).
It's likely that I'll add one more commit to this branch within the end of the week with the final changes.

cala commented:
Assigning myself to this.
I trust @evil-at-wow on the code part, I will look at the functional AH bot part.

@canidae This looks definitely interesting but as I have other CMaNGOS priorities to look at before, this might take a few time before I can finally review and merge this. I hope you don't mind. In the meantime, that will let you look at any balance or stability issue you may see.

👍

canidae commented:
There's one more change I've been considering (suggested in Discord):
Currently the AHBot use the same price both when selling and buying items. It might be an idea to be able to specify that the AHBot should buy for a certain percentage of what it would sell the same item for.
For example, if the AHBot would sell an item for 10G then maybe it should be possible to specify that it should buy the same item for 9G (90% of "item value") when put up on auction by a player.
This would make it more feasible to bid on items placed by the bot, and it makes it slightly more difficult to calculate the max price you can expect from an item.

My weekend is a bit busy so I don't know if I'll have the time before next week, so if you got more pressing matters to attend it's no problem :)

evil-at-wow commented:
> I trust @evil-at-wow on the code part, I will look at the functional AH bot part.

Just to rule out any potential misunderstanding: I didn't review the code changes of this PR. I just had a quick look here, saw the message about "strange segfaults" and commented on that.

If you think that's needed, I can try to do a more formal code review, but I'm not familiar with the AH bot code, so I'd have to find some time then before I'd be able to give meaningful feedback (I can always comment on "C++ syntax" in general to say so but that alone is typically not super useful).

canidae commented:
I wouldn't mind if someone could look over the code. I don't think deep knowledge of neither the AH nor AHBot code is needed. The existing code was a bit over engineered for my taste, but fairly easy to read, hopefully the new code is as easy to understand. Although, I must confess that building a "realistic" selection of items on the AH meant a bit more complicated code (and config).
As mentioned there's one more config option I'd like to add in, but I might not have time to finish it before next week. The addition will be quite small, though.

canidae commented:
I've added possibility to differentiate between what the AHBot will sell & buy items for. By default it'll pay 90% of the value it would sell the same item for (although there's some variance to value through the Value.Variance setting). The rationale is that bidding on items is more meaningful (AHBot won't practically always outbid you), and it makes sense from a game mechanics standpoint.

I also renamed "Price" to "Value" in settings as the former was a bit misleading.

Finally I made the setting for overriding value of vendor sold items a boolean instead of a percentage of vendor sell price (this is basically just a safety feature to prevent the possibility of players buying cheap uncommon or better BoE items from vendor and selling them at AH for great profit).

canidae commented:
If this PR is accepted, a change should be done in classic-db as well. `Full_DB/ClassicDB_1_10_1_z2742.sq` adds some commands to the command table that no longer will exist:
```
    ('ahbot items amount',3,'Syntax: .ahbot items amount $GreyItems $WhiteItems $GreenItems $BlueItems $PurpleItems $OrangeItems $YellowItems\r\n\r\nSet amount of each items color be selled on auction.'),
    ('ahbot items amount blue',3,'Syntax: .ahbot items amount blue $BlueItems\r\n\r\nSet amount of Blue color items be selled on auction.'),
    ('ahbot items amount green',3,'Syntax: .ahbot items amount green $GreenItems\r\n\r\nSet amount of Green color items be selled on auction.'),
    ('ahbot items amount grey',3,'Syntax: .ahbot items amount grey $GreyItems\r\n\r\nSet amount of Grey color items be selled on auction.'),
    ('ahbot items amount orange',3,'Syntax: .ahbot items amount orange $OrangeItems\r\n\r\nSet amount of Orange color items be selled on auction.'),
    ('ahbot items amount purple',3,'Syntax: .ahbot items amount purple $PurpleItems\r\n\r\nSet amount of Purple color items be selled on auction.'),
    ('ahbot items amount white',3,'Syntax: .ahbot items amount white $WhiteItems\r\n\r\nSet amount of White color items be selled on auction.'),
    ('ahbot items amount yellow',3,'Syntax: .ahbot items amount yellow $YellowItems\r\n\r\nSet amount of Yellow color items be selled on auction.'),
    ('ahbot items ratio',3,'Syntax: .ahbot items ratio $allianceratio $horderatio $neutralratio\r\n\r\nSet ratio of items in 3 auctions house.'),
    ('ahbot items ratio alliance',3,'Syntax: .ahbot items ratio alliance $allianceratio\r\n\r\nSet ratio of items in alliance auction house.'),
    ('ahbot items ratio horde',3,'Syntax: .ahbot items ratio horde $horderatio\r\n\r\nSet ratio of items in horde auction house.'),
    ('ahbot items ratio neutral',3,'Syntax: .ahbot items ratio neutral $neutralratio\r\n\r\nSet ratio of items in $neutral auction house.'),
```

cyberium commented:
I will test this, i already pointed in chat some concern about how this is configurable or not. It seem that fine tunning setting need more improvements. Maybe a way to override some prices or block some items.

I also asked if there is a possibility to have it fully optionnal like playerbot via cmake.

al3xc1985 commented:
+1 for implementing it like playerbot

canidae commented:
I will look into cmake. Meant to do it this week, but I've been exceedingly busy lately and haven't found the time yet.
As for item price tuning (do note that the current implementation has the same problem, so this is not a new flaw):
So far this implementation is as generic as possible, item value is determined by item quality, category and vendor sell/buy price (`AuctionHouseBot.Value.*` config). I've tried to set sensible values (basically, vendor sell price times 1/2/4/8 for white/green/blue/purple) and playing with this for some weeks now I'm _mostly_ satisfied:
An issue I've not found a good solution for is that skin/herb/ore all have a very low vendor sell price, and are thus quite cheap at the AH. It can't really be resolved by increasing sell price for trade goods as other trade goods, such as cloth, enchanting mats and more are reasonably valued, the price of these would then skyrocket.
I'm not a fan of making lots of intricate if's and switches, that just leads to unreadable code and probably other issues. Item value should be determined based on data available from the database, and I've not found a better way than using the vendor sell/buy price for this.
I have little faith in a "dynamic" economy where items are valued based on supply & demand. For one it is difficult to implement, it'll be CPU and/or RAM intensive, and quite likely easily abused by players. The AHBot is after all a work-around for servers with few players.

As for the possibility to add/block items, I've so far intentionally left it out (but I don't oppose adding it back in). One of the reason being that all items you naturally can loot/craft in the game (and that can be put up on the AH) should have a chance to appear at the AH (unless I've overlooked some item source). Another is that you'll likely need a complex configuration to proper control how items are added to the AH. For each item you'll likely want to control the chance of it being added, the stack size and possibly the item value. I'm not entirely convinced this will be an improvement.

I'd gladly improve it further so feel free to come up with suggestions. The price of certain trade goods is bothering me, any ideas here is most welcome.

canidae commented:
I have one remaining idea to solve the issue with underpriced/overpriced items (which coincidentally adds some dynamic to AH):
* When AHBot sells/buys an item, the value of that item is increased/decreased slightly.

Con:
* Item value needs to be persisted through server restarts, most likely a new database table is needed, I'm guessing this belongs to the realm database.
* This opens up for AH manipulation and might make it possible for players to easily have a source of infinite income.
* Adds more complexity to the code.
* It'll take time before prices "stabilize" (when players would rather farm items than buy of the AH). Fish will notably take a long time as they're sold for coppers from vendor, but can be worth gold at the AH.

I'd like to hear the opinion of the devs on this, it's not something I'll spend time on unless they think this will be an improvement over static item values.

al3xc1985 commented:
How about adding a db table for items to persist over restarts, and add a new option especially to modify hebs/mine/skinning/etc like price. In that db u can give them a vollumn for examole type +herb, mine, skinning,etc) and motify price accordingly to this option

al3xc1985 commented:
U cand also have that option to modify all prices depending on the items type for example

canidae commented:
We discussed a similar solution at Discord, and I'm looking into this. Will take some time as I'm quite busy these days.

canidae commented:
I'm not actively into CManGOS anymore, but I still believe this implementation is an improvement over the existing one. I'd like this to be my contribution to the project and as of such I'll gladly be of assistance for this implementation in the future if needed.

cala commented:
> I'm not actively into CManGOS anymore, but I still believe this implementation is an improvement over the existing one. I'd like this to be my contribution to the project and as of such I'll gladly be of assistance for this implementation in the future if needed.

Greetings @canidae !

Again, thank you very much for this PR and the time and work you've put in it.

I'm currently in the process of merging PRs and closing issues that were long delayed due to the (unexpectedly) long time I spent on Naxxramas rework. I definitely want to merge your work but it still will have to wait a few days possibly weeks more, the time that I empty the pipe of PR/issues/ToDo things I'm dealing with.

I hope you'll still be available if I have questions when I shall work on this PR. 😄

canidae commented:
Heya cala. Just drop me a line when you have the time. Until then I won't bother resolving the conflicts caused by changes in upstream. Did it once already and it didn't take many days before the database changed again (and changes to the database cause a conflict). I can fix it up when you'd like to review this.

cala commented:
I test this PR for a few days and so far, so good regarding the AH bot seller. I am performing a few more tests on the AH bot buyer before merging. 👍

cala commented on src/shared/Log.h:
```diff
@@ -52,15 +52,13 @@ enum LogFilters
-    LOG_FILTER_MAP_LOADING        = 0x020000,               // 17 Map loading/unloading (MAP, VMAPS, MMAP)
-    LOG_FILTER_EVENT_AI_DEV       = 0x040000,               // 18 Event AI actions
-    LOG_FILTER_DB_SCRIPT          = 0x080000,               // 19 Db scripts
+    LOG_FILTER_PATHFINDING        = 0x010000,               // 14 Pathfinding
```
I'd like to have @cyberium or @killerwife opinion here about the change in the way logs are indexed. Changing the indexes on the right without changing the enum on the left can be misleading, in my opinion. If the enums were left untouched on purpose, then the indexing should not be changed and the removed ones should be marked as free for re-use.

canidae replied:
Given that there already is `// reserved for future version` for a "missing" enum value I should've done similar for those I removed and not messed with the index numbers (not even sure why I did that).
I'll fix this when I can spare the time. Good catch.

canidae replied:
Might even be an error here. Noticed just now that LOG_FILTER_COUNT seems to include the skipped 3rd index. I skipped two additional indexes, should then LOG_FILTER_COUNT remain 20?
I'll have to dig deeper into the code to understand how this constant is used.

cala replied:
Aside from this possible logging issue, the PR seems fine to me and I shall merge it once the logs are sorted out. 👍

canidae replied:
That's great, thanks for taking the time!
I think I might be able to look at the log issue after work (in 4-5 hours). I'll get back to you then.

canidae replied:
Looking through the code I don't think there was an actual error, but as noted it was misleading to rearrange the indexes like I did.
I've now added some empty entries for the freed log filters so the slots can be reused in the future.

cala commented:
Merged. Thank you very much, nice work!
